(un)linkapps: modernize, prune: remove broken app symlinks

UniqMartin opened the pull request:
*— Replaces Homebrew/legacy-homebrew#46549 as it no longer cleanly applies without a rebase.*

Summary of changes:

- Re-implemented `linkapps` and `unlinkapps` to use `Pathname` and other Ruby-native facilities.
- Added `--dry-run` option to `unlinkapps`.
- Extended `prune` to delete stale `.app` symlinks created by `linkapps`.
- Adapted documentation and Bash completion to reflect changes.
- Adjusted tests to account for the slightly changed behavior and output of modified commands.

Out of scope, but planned for a future PR:

- Auto-pruning of stale `.app` symlinks on `uninstall`, `upgrade`, … (see, e.g., Homebrew/legacy-homebrew#46542 and possibly others).
- Additional refactoring of shared functionality to avoid cross-`cmd/` includes.

----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] ~~Have you written new tests for your changes?~~
- [x] Have you successfully ran `brew tests` with your changes locally?

apjanke commented:
I like the new code. Seems more maintainable to me with the native Ruby stuff.

Looks like the failure is either an output format change, or failure to respect `--local`.

```
# Running:

.........F..................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 71.055511s, 11.8780 runs/s, 28.6959 assertions/s.

  1) Failure:
IntegrationCommandTests#test_prune [/usr/local/Library/Homebrew/test/test_integration_cmds.rb:697]:
Expected /Nothing\ pruned\nNo\ apps\ unlinked\ from\ \/Applications\nNo\ apps\ unlinked\ from\ \/Users\//
 to match "Nothing pruned\nNo apps unlinked from /Applications"
.

844 runs, 2039 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
```

mikemcquaid commented on Library/Homebrew/cmd/linkapps.rb:
```diff
@@ -1,12 +1,11 @@
+    target_dir = linkapps_target(:local => ARGV.include?("--local"))
 
-    unless File.exist? target_dir
+    unless target_dir.directory?
```
Personally I dislike `unless`/`else` as I find it hard to parse. Feel free to ignore me though: not your code.

UniqMartin replied:
Agreed, but the `else` you're seeing belongs to the `if` following the `unless`. This `unless` block has no `else` branch. :wink:

mikemcquaid commented on Library/Homebrew/cmd/linkapps.rb:
```diff
@@ -16,24 +15,40 @@ def linkapps
         end
 
-        system "ln", "-sf", app, target_dir
+        # We cannot use `install_symlink` because we want an absolute link.
```
I think it may be worth using the system call. I forget why but I vaguely remember `FileUtils.ln_sf` having weird failure conditions hence the `system` use. Yes, that should be a comment 😛

mikemcquaid commented:
Looks good to me except the test failure.

UniqMartin commented:
Addressed the comments and (hopefully) fixed the test failure.

> I like the new code. Seems more maintainable to me with the native Ruby stuff.

Thanks!

> Looks like the failure is either an output format change, or failure to respect `--local`.

That was just my somewhat naive assumption that the user we're testing with by default has an `Applications` directory in their home directory. I have now relaxed the test to only assume that `/Applications` exists and that should get rid of the failure.

mikemcquaid commented:
LGTM!
